types.go

```diff
@@ -0,0 +1,160 @@
+// Copyright (c) 2016 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package tally
+
+import "time"
+
+// Scope is a namespace wrapper around a stats reporter, ensuring that
+// all emitted values have a given prefix or set of tags
+type Scope interface {
+	// Counter returns the Counter object corresponding to the name
+	Counter(name string) Counter
+
+	// Gauge returns the Gauge object corresponding to the name
+	Gauge(name string) Gauge
+
+	// Timer returns the Timer object corresponding to the name
+	Timer(name string) Timer
+
+	// Tagged returns a new scope with the given tags
+	Tagged(tags map[string]string) Scope
+
+	// SubScope returns a new child scope with the given name
+	SubScope(name string) Scope
```

Comment:
these two methods, SubScope and Tagged, remind me of another difficulty I always had with go-common API w.r.t. proper separation of metric name and tags. Case in point - look at the [Metrics](https://github.com/uber/jaeger-client-go/blob/master/metrics.go#L31) struct in Jaeger client, especially the counters with tag `group` - I don't have a good feeling whether this should be a tag (dimension) or just a part of the metric name/namespace. It would be good to provide concrete examples when and what should be used in the comments. 

In particular, this is different from say go-kit where each metric type has a `WithTags()` method. That usage is more clear to me - you define a metric with a certain fixed name (and namespace), inject that metric into some service, and then the service has the option of adding dimensions.

This is a very subtle aspect of this API design, where a good design might clarify a lot of things, and a bad design makes the intended usage confusing.